File: Students.js
const prompt = require('prompt-sync')();

class Student {
  constructor(username, name, age, city, zip, courses) {
    this.username = username;
    this.name = name;
    this.age = age;
    this.city = city;
    this.zip = zip;
    this.courses = courses;
}
}

const students = [];
const coursesOffered = [
  "Math", "Physics", "Computer Science", "Biology", "Chemistry",
  "Statistics", "English", "Economics", "History", "Philosophy",
  "Sociology", "Political Science", "Geography", "Psychology", "Art",
  "Music", "Engineering", "Law", "Medicine", "Business"
];

function addStudent() {
  const username = prompt("Enter username: ");
  const name = prompt("Enter name: ");
  const age = parseInt(prompt("Enter age: "));
  const city = prompt("Enter city: ");
  const zip = prompt("Enter zip code: ");
  const courseInput = prompt("Enter courses (comma-separated): ");
  const courses = courseInput.split(",").map(c => c.trim());
  const newStudent = new Student(username, name, age, city, zip, courses);
  students.push(newStudent);
  console.log("Student added successfully!");
}

function findStudent(username) {
  return students.find(student => student.username === username);
}

function displayStudent() {
  const username = prompt("Enter username to display: ");
  const student = findStudent(username);
  if (student) {
    console.log(`Name: ${student.name}, Age: ${student.age}, City: ${student.city}, Zip: ${student.zip}, Courses: ${student.courses.join(", ")}`);
} else {
    console.log("Student not found.");
}
}

function addCourse() {
  const username = prompt("Enter username: ");
  const course = prompt("Enter course to add: ");
  const student = findStudent(username);
  if (student) {
    if (!coursesOffered.includes(course)) {
      console.log("Course not offered.");
} else if (student.courses.includes(course)) {
      console.log("Course already exists.");
} else {
      student.courses.push(course);
      console.log(`${course} added successfully!`);
}
} else {
    console.log("Student not found.");
}
}

function updateCourse() {
  const username = prompt("Enter username: ");
  const oldCourse = prompt("Enter course to replace: ");
  const newCourse = prompt("Enter new course: ");
  const student = findStudent(username);
  if (student) {
    const index = student.courses.indexOf(oldCourse);
    if (index !== -1 && coursesOffered.includes(newCourse)) {
      student.courses[index] = newCourse;
      console.log(`${oldCourse} updated to ${newCourse}`);
} else {
      console.log("Invalid course update.");
}
} else {
    console.log("Student not found.");
}
}

function updateStudent() {
  const username = prompt("Enter username: ");
  const student = findStudent(username);
  if (student) {
    const name = prompt("Enter new name (or leave blank): ");
    const ageInput = prompt("Enter new age (or leave blank): ");
    const city = prompt("Enter new city (or leave blank): ");
    const zip = prompt("Enter new zip (or leave blank): ");

    if (name) student.name = name;
    if (ageInput) student.age = parseInt(ageInput);
    if (city) student.city = city;
    if (zip) student.zip = zip;

    console.log("Student updated!");
} else {
    console.log("Student not found.");
}
}

function totalStudents() {
  console.log(`Total students: ${students.length}`);
}


while (true) {
  console.log("\n--- Student Management System ---");
  console.log("1. Add Student");
  console.log("2. Display Student");
  console.log("3. Add Course");
  console.log("4. Update Course");
  console.log("5. Update Student");
  console.log("6. Total Students");
  console.log("7. Exit");

  const choice = prompt("Choose an option: ");
  switch (choice) {
    case "1": addStudent(); break;
    case "2": displayStudent(); break;
    case "3": addCourse(); break;
    case "4": updateCourse(); break;
    case "5": updateStudent(); break;
    case "6": totalStudents(); break;
    case "7": console.log("Goodbye!"); process.exit();
    default: console.log("Invalid choice.");
}
}

File: Students.java
import java.util.Arrays;

class Student {
String username;
String name;
int age;
String city;
String zip;
String[] courses;
}

public class Students{
static Student[] students = new Student[100];
static int totalStudents = 0;

static String[] coursesOffered = {
        "Math","Physics","Computer Science","Biology","Chemistry",
        "Statistics","English","Economics","History","Philosophy",
        "Sociology","Political Science","Geography","Psychology","Art",
        "Music","Engineering","Law","Medicine","Business"
};

public static void addStudent(String username, String name, int age, String city, String zip, String[] courses) {
        Student newStudent = new Student();
        newStudent.username = username;
        newStudent.name = name;
        newStudent.age = age;
        newStudent.city = city;
        newStudent.zip = zip;
        newStudent.courses = courses;
        students[totalStudents] = newStudent;
        totalStudents++;
    }

public static Student findStudent(String username) {
for (int counter = 0; counter < totalStudents; counter++) {
if (students[counter].username.equals(username)) {
                return students[counter];
}
}
        return null;
}

public static void displayStudent(String username) {
Student studentRecord = findStudent(username);
if (studentRecord != null) {
System.out.println("Name: " + studentRecord.name +
                               ", Age: " + studentRecord.age +
                               ", City: " + studentRecord.city +
                               ", Zip: " + studentRecord.zip +
                               ", Courses: " + Arrays.toString(studentRecord.courses));
}
}

public static void displayCourses(String username) {
Student studentRecord = findStudent(username);
if (studentRecord != null) {
System.out.println(Arrays.toString(studentRecord.courses));
}
}

public static void displayZip(String username) {
Student studentRecord = findStudent(username);
if (studentRecord != null) {
System.out.println(studentRecord.zip);
}
}

public static void displayCity(String username) {
Student studentRecord = findStudent(username);
if (studentRecord != null) {
System.out.println(studentRecord.city);
}
}

public static void addCourse(String username, String course) {
Student studentRecord = findStudent(username);
if (studentRecord != null) {
boolean offered = false;
for (int counter = 0; counter < coursesOffered.length; counter++) {
if (coursesOffered[counter].equals(course)) {
                    offered = true;
}
}
boolean exists = false;
for (int counter = 0; counter < studentRecord.courses.length; counter++) {
if (studentRecord.courses[counter].equals(course)) {
                    exists = true;
}
}
if (offered && !exists) {
String[] updatedCourses = Arrays.copyOf(studentRecord.courses, studentRecord.courses.length + 1);
updatedCourses[studentRecord.courses.length] = course;
studentRecord.courses = updatedCourses;
System.out.println(course + " added successfully!");
} else {
System.out.println("Invalid or duplicate course!");
}
}
}

public static void updateCourse(String username, String oldCourse, String newCourse) {
Student studentRecord = findStudent(username);
if (studentRecord != null) {
for (int counter = 0; counter < studentRecord.courses.length; counter++) {
if (studentRecord.courses[counter].equals(oldCourse)) {
boolean offered = false;
for (int index = 0; index < coursesOffered.length; index++) {
if (coursesOffered[index].equals(newCourse)) {
                   offered = true;
}
}
if (offered) {
studentRecord.courses[counter] = newCourse;
System.out.println(oldCourse + " updated to " + newCourse);
} else {
System.out.println("New course not offered!");
}
        return;
}
}
System.out.println("Old course not found!");
}
}

public static void updateStudent(String username, String name, Integer age, String city, String zip) {
Student studentRecord = findStudent(username);
if (studentRecord != null) {
if (name != null) studentRecord.name = name;
if (age != null) studentRecord.age = age;
if (city != null) studentRecord.city = city;
if (zip != null) studentRecord.zip = zip;
System.out.println("Student updated!");
}
}

public static void totalStudents() {
System.out.println("Total students: " + totalStudents);
}

public static void main(String[] args) {
        addStudent("john22", "John Paul", 22, "Lagos", "101001", new String[]{"Math", "Physics"});
        displayStudent("john22");
        displayCourses("john22");
        displayZip("john22");
        displayCity("john22");
        addCourse("john22", "Computer Science");
        updateCourse("john22", "Math", "Biology");
        updateStudent("john22", null, 23, "Abuja", null);
        totalStudents();
}
}